// jexia-sdk-js/_esm5/api/fileops/fileOperationsModule.js
import { RequestExecuter } from "../../internal/executer";
import { ClientConfiguration } from "../core/client";
import { AuthOptions } from "../core/tokenManager";
import { FilesetName } from "./fileops.interfaces";
import { Fileset } from "./fileset";
import { FileUploader } from "./fileUploader";
var defaultConfiguration = {
    uploadWaitForCompleted: false,
    uploadTimeout: 120000
};
var FileOperationsModule = (function () {
    function FileOperationsModule(formData, config) {
        this.formData = formData;
        this.config = Object.assign(defaultConfiguration, config);
    }
    FileOperationsModule.prototype.init = function (coreInjector) {
        this.injector = coreInjector.resolveAndCreateChild([
            RequestExecuter,
        ]);
        if (this.config.uploadWaitForCompleted) {
            var isRTCModuleActive = Boolean(this.injector.get(ClientConfiguration).rtc);
            if (!isRTCModuleActive) {
                return Promise.reject("RTC module needs to be activated for automatic file uploading subscription");
            }
        }
        return Promise.resolve(this);
    };
    FileOperationsModule.prototype.getConfig = function () {
        return { fileOperations: this.config };
    };
    FileOperationsModule.prototype.fileset = function (fileset, auth) {
        var config = this.injector.get(AuthOptions);
        if (auth) {
            config.auth = auth;
        }
        var injector = this.injector.resolveAndCreateChild([
            {
                provide: FilesetName,
                useValue: fileset,
            },
            {
                provide: AuthOptions,
                useValue: config,
            },
            FileUploader,
            Fileset,
        ]);
        injector.get(FileUploader).provideFormData(this.formData);
        return injector.get(Fileset);
    };
    FileOperationsModule.prototype.filesets = function (filesets, auth) {
        var _this = this;
        return filesets.map(function (fileset) { return _this.fileset(fileset, auth); });
    };
    FileOperationsModule.prototype.terminate = function () {
        return Promise.resolve(this);
    };
    return FileOperationsModule;
}());
export { FileOperationsModule };
