// jexia-sdk-js/_esm5/api/fileops/fileUploader.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Inject, Injectable } from "injection-js";
import { merge } from "rxjs";
import { map, switchMap } from "rxjs/operators";
import { API } from "../../config";
import { RequestAdapter } from "../../internal/requestAdapter";
import { AuthOptions, TokenManager } from "../core/tokenManager";
import { FilesetName } from "./fileops.interfaces";
var FileUploader = (function () {
    function FileUploader(config, filesetName, tokenManager, requestAdapter) {
        this.config = config;
        this.filesetName = filesetName;
        this.tokenManager = tokenManager;
        this.requestAdapter = requestAdapter;
    }
    FileUploader.prototype.provideFormData = function (formData) {
        this.formData = this.resetFormData(formData);
    };
    FileUploader.prototype.upload = function (files) {
        var _this = this;
        return merge.apply(void 0, files.map(function (file) { return _this.uploadFile(file); }));
    };
    FileUploader.prototype.uploadFile = function (record) {
        var _this = this;
        var formData = this.resetFormData();
        formData.append("data", record.data ? JSON.stringify(record.data) : "{}");
        if (record.file) {
            formData.append("file", record.file);
        }
        return this.tokenManager.token(this.config.auth).pipe(switchMap(function (token) { return _this.execute(token, formData); }), map(function (_a) {
            var uploaded = _a[0];
            return uploaded;
        }));
    };
    FileUploader.prototype.execute = function (token, formData) {
        var headers = {
            Authorization: "Bearer " + token,
        };
        if (formData.getHeaders) {
            Object.assign(headers, formData.getHeaders());
        }
        return this.requestAdapter.upload(this.getUrl(), headers, formData);
    };
    ;
    FileUploader.prototype.getUrl = function () {
        return [
            API.PROTOCOL + "://" + this.config.projectID + "." + API.HOST + "." + API.DOMAIN + ":" + API.PORT,
            API.FILES.ENDPOINT,
            this.filesetName,
        ].join("/");
    };
    FileUploader.prototype.resetFormData = function (formData) {
        if (formData === void 0) { formData = this.formData; }
        return new formData.constructor();
    };
    FileUploader = __decorate([
        Injectable(),
        __param(0, Inject(AuthOptions)),
        __param(1, Inject(FilesetName)),
        __metadata("design:paramtypes", [Object, String, TokenManager,
            RequestAdapter])
    ], FileUploader);
    return FileUploader;
}());
export { FileUploader };
