// jexia-sdk-js/_esm5/api/fileops/fileset.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Inject, Injectable } from "injection-js";
import { Observable, Subject } from "rxjs";
import { concatMap, filter, map, shareReplay, takeUntil, tap } from "rxjs/operators";
import { RequestExecuter } from "../../internal/executer";
import { QueryActionType } from "../../internal/utils";
import { ClientConfiguration } from "../core/client";
import { ActionQuery } from "../core/queries/actionQuery";
import { DeleteQuery } from "../core/queries/deleteQuery";
import { SelectQuery } from "../core/queries/selectQuery";
import { UpdateQuery } from "../core/queries/updateQuery";
import { ResourceType } from "../core/resource";
import { FilesetName, IFileStatus } from "../fileops/fileops.interfaces";
import { FileUploader } from "../fileops/fileUploader";
var Fileset = (function () {
    function Fileset(filesetName, clientConfig, fileUploader, requestExecuter) {
        this.filesetName = filesetName;
        this.clientConfig = clientConfig;
        this.fileUploader = fileUploader;
        this.requestExecuter = requestExecuter;
        this.resourceType = ResourceType.Fileset;
    }
    Object.defineProperty(Fileset.prototype, "name", {
        get: function () {
            return this.filesetName;
        },
        enumerable: true,
        configurable: true
    });
    Fileset.prototype.upload = function (files) {
        var fileUploadObservable = this.fileUploader.upload(files);
        if (this.clientConfig.fileOperations.uploadWaitForCompleted) {
            return this.getFileUpdates(fileUploadObservable, files.length);
        }
        return fileUploadObservable;
    };
    Fileset.prototype.select = function () {
        return new SelectQuery(this.requestExecuter, ResourceType.Fileset, this.filesetName);
    };
    Fileset.prototype.update = function (data) {
        return new UpdateQuery(this.requestExecuter, data, ResourceType.Fileset, this.filesetName);
    };
    Fileset.prototype.insert = function () {
        throw new Error("Fileset does not have INSERT, please use upload() instead");
    };
    Fileset.prototype.delete = function () {
        return new DeleteQuery(this.requestExecuter, ResourceType.Fileset, this.filesetName);
    };
    Fileset.prototype.attach = function (resourceName, filter) {
        return new ActionQuery(this.requestExecuter, this.resourceType, this.filesetName, resourceName, QueryActionType.ATTACH, filter);
    };
    Fileset.prototype.detach = function (resourceName, filter) {
        return new ActionQuery(this.requestExecuter, this.resourceType, this.filesetName, resourceName, QueryActionType.DETACH, filter);
    };
    Fileset.prototype.getFileUpdates = function (uploadingProcess, filesUploaded) {
        var filesCompleted = 0;
        var allFilesCompleted = new Subject();
        var sharedUploadingProcess = uploadingProcess.pipe(shareReplay());
        var watcher = this.watch().pipe(takeUntil(allFilesCompleted), filter(function (event) { return event.action === "updated"; }), concatMap(function (event) { return sharedUploadingProcess.pipe(filter(function (fileRecord) { return fileRecord.id === event.data[0].id; }), map(function (fileRecord) {
            fileRecord.status = IFileStatus.COMPLETED;
            filesCompleted++;
            return fileRecord;
        })); }), tap(function () {
            if (filesCompleted === filesUploaded) {
                allFilesCompleted.next();
                allFilesCompleted.complete();
            }
        }));
        return Observable.create(function (subscriber) {
            var watcherSubscription = watcher.subscribe(subscriber);
            sharedUploadingProcess.subscribe();
            return function () { return watcherSubscription.unsubscribe(); };
        });
    };
    Fileset = __decorate([
        Injectable(),
        __param(0, Inject(FilesetName)),
        __param(1, Inject(ClientConfiguration)),
        __metadata("design:paramtypes", [String, Object, FileUploader,
            RequestExecuter])
    ], Fileset);
    return Fileset;
}());
export { Fileset };
