// jexia-sdk-js/_esm5/api/realtime/channel.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { from, Observable } from "rxjs";
import { RequestExecuter } from "../../internal/executer";
import { Query } from "../../internal/query";
import { RequestMethod } from "../../internal/requestAdapter.interfaces";
import { ResourceType } from "../core/resource";
import { RealTimeCommandTypes } from "./realTime.interfaces";
import { realTimeCommand, subscribeEventMessage, unsubscribeEventMessage, wsReadyDefer } from "./websocket";
var Channel = (function (_super) {
    __extends(Channel, _super);
    function Channel(injector, websocketFactory, name) {
        var _this = _super.call(this, function (observer) {
            wsReadyDefer.promise
                .then(function () { return subscribeEventMessage(_this.websocketFactory(), ["published"], name, ResourceType.Channel, observer); })
                .catch(function (error) { return observer.error(error); });
            return function () { return wsReadyDefer.promise
                .then(function () { return unsubscribeEventMessage(_this.websocketFactory(), ["published"], name, ResourceType.Channel, observer)
                .catch(function () { return undefined; }); }); };
        }) || this;
        _this.injector = injector;
        _this.websocketFactory = websocketFactory;
        _this.name = name;
        return _this;
    }
    Channel.prototype.publish = function (data) {
        var _this = this;
        return from(wsReadyDefer.promise.then(function () {
            return realTimeCommand(_this.websocketFactory(), {
                command: RealTimeCommandTypes.Publish,
                correlation_id: Math.random().toString(),
                arguments: { channel: _this.name, data: data },
            });
        }));
    };
    Channel.prototype.getLog = function (filter) {
        var query = new Query();
        if (filter) {
            query.setFilterCriteria(filter);
        }
        var requestExecutor = this.injector.get(RequestExecuter);
        return requestExecutor.executeRequest({
            resourceType: ResourceType.Channel,
            resourceName: this.name,
            method: RequestMethod.GET,
            body: {},
            queryParams: query.compileToQueryParams()
        });
    };
    return Channel;
}(Observable));
export { Channel };
