// jexia-sdk-js/_esm5/api/core/filteringApi.js
import { CompositeFilteringCondition, FilteringCondition } from "./filteringCondition";
var FieldFilter = (function () {
    function FieldFilter(fieldName) {
        this.fieldName = fieldName;
    }
    FieldFilter.prototype.isGreaterThan = function (value) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, ">", value));
    };
    FieldFilter.prototype.isLessThan = function (value) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "<", value));
    };
    FieldFilter.prototype.isEqualTo = function (value) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "=", value));
    };
    FieldFilter.prototype.isDifferentFrom = function (value) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "!=", value));
    };
    FieldFilter.prototype.isEqualOrGreaterThan = function (value) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, ">=", value));
    };
    FieldFilter.prototype.isEqualOrLessThan = function (value) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "<=", value));
    };
    FieldFilter.prototype.isNull = function () {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "null", true));
    };
    FieldFilter.prototype.isNotNull = function () {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "null", false));
    };
    FieldFilter.prototype.isInArray = function (values) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "in", values));
    };
    FieldFilter.prototype.isNotInArray = function (values) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "not in", values));
    };
    FieldFilter.prototype.isLike = function (value) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "like", value));
    };
    FieldFilter.prototype.satisfiesRegexp = function (regexp) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "regex", regexp));
    };
    FieldFilter.prototype.isBetween = function (start, end) {
        return new FilteringCriterion(new FilteringCondition(this.fieldName, "between", [start, end]));
    };
    return FieldFilter;
}());
export { FieldFilter };
var FilteringCriterion = (function () {
    function FilteringCriterion(lowLevelCondition, highLevelCriteria) {
        if (lowLevelCondition) {
            this.lowLevelCondition = lowLevelCondition;
        }
        if (highLevelCriteria) {
            this.lowLevelCondition = new CompositeFilteringCondition(highLevelCriteria.lowLevelCondition, "and");
        }
        if (!this.lowLevelCondition) {
            throw new Error("No information was given when constructing a FilteringCriterion.");
        }
    }
    Object.defineProperty(FilteringCriterion.prototype, "condition", {
        get: function () {
            return this.lowLevelCondition;
        },
        enumerable: true,
        configurable: true
    });
    FilteringCriterion.prototype.and = function (conditionToAdd) {
        return new FilteringCriterion(this.lowLevelCondition.and(conditionToAdd.lowLevelCondition));
    };
    FilteringCriterion.prototype.or = function (conditionToAdd) {
        return new FilteringCriterion(this.lowLevelCondition.or(conditionToAdd.lowLevelCondition));
    };
    return FilteringCriterion;
}());
export { FilteringCriterion };
export function field(name) {
    return new FieldFilter(name);
}
export function combineCriteria(criteria) {
    return new FilteringCriterion(undefined, criteria);
}
export function toFilteringCriterion(filter) {
    return typeof filter === "function"
        ? filter(function (f) { return new FieldFilter(f); })
        : filter;
}
