// jexia-sdk-js/internal/executer.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
var injection_js_1 = require("injection-js");
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var client_1 = require("../api/core/client");
var resource_1 = require("../api/core/resource");
var tokenManager_1 = require("../api/core/tokenManager");
var config_1 = require("../config");
var requestAdapter_1 = require("./requestAdapter");
var resourceEndpoints = (_a = {},
    _a[resource_1.ResourceType.Dataset] = config_1.API.DATA.ENDPOINT,
    _a[resource_1.ResourceType.Fileset] = config_1.API.FILES.ENDPOINT,
    _a[resource_1.ResourceType.Channel] = config_1.API.CHANNEL.ENDPOINT,
    _a[resource_1.ResourceType.Users] = config_1.API.UMS.ENDPOINT,
    _a);
var RequestExecuter = (function () {
    function RequestExecuter(config, systemInit, requestAdapter, tokenManager) {
        this.config = config;
        this.systemInit = systemInit;
        this.requestAdapter = requestAdapter;
        this.tokenManager = tokenManager;
    }
    RequestExecuter.prototype.executeRequest = function (request) {
        var _this = this;
        var URI = this.getUrl(request) + this.parseQueryParams(request);
        return rxjs_1.combineLatest([
            rxjs_1.from(this.systemInit),
            this.getRequestOptions(request),
        ]).pipe(operators_1.switchMap(function (_a) {
            var options = _a[1];
            return _this.requestAdapter.execute(URI, options);
        }));
    };
    RequestExecuter.prototype.getRequestOptions = function (request) {
        var _this = this;
        return this.tokenManager.token(this.config.auth).pipe(operators_1.map(function (token) {
            var options = {
                headers: { Authorization: "Bearer " + token },
                method: request.method,
            };
            if (_this.hasBody(request)) {
                options.body = request.body;
            }
            return options;
        }));
    };
    RequestExecuter.prototype.hasBody = function (_a) {
        var method = _a.method;
        return ![requestAdapter_1.RequestMethod.GET, requestAdapter_1.RequestMethod.DELETE].includes(method);
    };
    RequestExecuter.prototype.parseQueryParams = function (_a) {
        var _b = _a.queryParams, queryParams = _b === void 0 ? [] : _b;
        if (!queryParams.length) {
            return "";
        }
        var encodeValue = function (v) { return encodeURIComponent(typeof v === "string" ? v : JSON.stringify(v)); };
        var toQueryParam = function (_a) {
            var key = _a.key, value = _a.value;
            return key + "=" + encodeValue(value);
        };
        return "?" + queryParams
            .map(toQueryParam)
            .join("&");
    };
    RequestExecuter.prototype.getUrl = function (_a) {
        var resourceType = _a.resourceType, resourceName = _a.resourceName;
        var endpoint = resourceEndpoints[resourceType];
        return [
            config_1.API.PROTOCOL + "://" + this.config.projectID + "." + config_1.API.HOST + "." + config_1.API.DOMAIN + ":" + config_1.API.PORT,
            endpoint,
            resourceName,
        ].join("/");
    };
    RequestExecuter = __decorate([
        injection_js_1.Injectable(),
        __param(0, injection_js_1.Inject(tokenManager_1.AuthOptions)),
        __param(1, injection_js_1.Inject(client_1.ClientInit)),
        __metadata("design:paramtypes", [Object, Promise, requestAdapter_1.RequestAdapter,
            tokenManager_1.TokenManager])
    ], RequestExecuter);
    return RequestExecuter;
}());
exports.RequestExecuter = RequestExecuter;
